Automattic/woocommerce-payments
# Tool to update Multi-Currency meta data

Fixes https://github.com/Automattic/woocommerce-payments-server/issues/3431

#### Changes proposed in this Pull Request

This PR adds a tool that, when enabled, will allow a store admin to modify/add Multi-Currency exchange rate meta data in an order. This helps to fix issues where we've seen this data missing on orders, which causes incorrect values in Analytics reports.

* Add new `MultiCurrency/Helpers` directory for this and any additional helpers.
* Add new `OrderMetaHelper` class for the tool.
* Add `OrderMetaHelper.md` to describe usage of the tool.
* Added new flag and method `WC_Payments_Features` to allow for enabling and disabling the helper.

The helper itself adds a new meta box to the order. In that meta box it displays information about the store, order, and intent:

* Store currency
* Order currency
* Order default currency (store currency at the time of order)
* Payment method ID
* Payment method title
* Intent ID
* Intent currency (from order meta)
* Multi-Currency exchange rate
* Stripe exchange rate (from order meta)
* Charge exchange rate (from intent)
* Charge currency (from intent)

The list is extensive to try to provide a full picture to the merchant, but it can be narrowed down easily.

A single field is added in the meta box to allow the merchant to enter a new exchange rate for the order. Once the field is populated and the order is saved, the meta data is added/updated. An order note is created under the user making the change, and new meta data is added in order to log the changes.

The meta box will not show on orders in the default currency, however, it will show for any order that is _not_ through WooCommerce Payments. This is due to Multi-Currency works for WooCommerce and only relies on WCPay for getting the automatic rates. Customers are able to purchase items with any gateway if the feature is enabled. For this reason, we only update one of the two meta values that hold exchange rates.

#### Testing instructions

Setup:
* Go to WooCommerce > Settings > Multi-Currency and enable at least one other currency in your store.

Testing:
* Purchase a product in another currency other than the default currency in your store.
* Go to the order under WooCommerce > Orders in the admin.
* Confirm there is not a meta box named _Multi-Currency Meta Helper_ showing.
* Go to WooCommerce > Status > Tools
* Locate the _Enable/Disable Multi-Currency Meta Helper_ row and click the _Enable/Disable_ button to enable the tool.
* Confirm the "Multi-Currency Meta Helper has been enabled" message shows.
* Return to your new order and confirm that the meta box named _Multi-Currency Meta Helper_ now shows.
* Delete both the `_wcpay_multi_currency_stripe_exchange_rate` and `_wcpay_multi_currency_order_exchange_rate` meta items under the order.
  * This can be done by adding this snippet somewhere, which enables viewing hidden meta data: `add_filter( 'is_protected_meta', '__return_false', 999 );`, or
  * This can be done in the `wp_postmeta` table by searching for your order number under `post_id`.
* Refresh the order.
* The meta box should now show:
  *  _Multi-Currency Exchange Rate_ is _Not found_.
  *  _Stripe Exchange Rate_ is _Not found_.
  *  If the intent is valid, there should be a _Suggested exchange rate_ with the _Charge exchange rate_ value shown.
* Enter text into the field and hit enter/return or click the _Update_ button for the order.
* An error should show that non-numeric information was passed: _Error: Exchange rate value not numeric: xxx_
* Copy and paste the suggest rate into the box and hit enter/return or click the _Update_ button for the order.
* A new note should show in the order notes under your user ID.
* The note should state the exchange rate was updated and include the from and to rates.
* The _Multi-Currency Exchange Rate_ value should now be populated in the meta box, the _Stripe Exchange Rate_ value will not be.
* A new meta entry under the order with the `meta_key` will be added. It will be a multidimensional array the key being the timestamp, and the value being an array with your user ID, the old rate, and the new rate. Each update to the exchange rate will add another entry in this log.
  * This log will need to be viewed directly in the database, or with a plugin like [Post Meta Inspector](https://wordpress.org/plugins/post-meta-inspector/).
* Place an order for a product using the default store currency.
* View the order under WooCommerce > Orders in the admin.
* The meta box should not show.
* Go back to WooCommerce > Status > Tools.
* Locate the _Enable/Disable Multi-Currency Meta Helper_ row and click the _Enable/Disable_ button to disable the tool.
* Confirm the "Multi-Currency Meta Helper has been disabled" message shows.
* Return to the order where you updated the exchange rate.
* Confirm the meta box no longer shows.


-------------------


**Post merge**
